pepper-apis/src/main/java/org/broadinstitute/ddp/route/DeleteActivityInstanceRoute.java

```diff
@@ -53,16 +53,19 @@ public ActivityInstanceSummary handle(Request request, Response response) {
             ActivityDefStore activityStore = ActivityDefStore.getInstance();
             FormActivityDef activityDef = ActivityInstanceUtil.getActivityDef(handle, activityStore, instanceDto, studyGuid);
 
+            Long previousInstanceId = ActivityInstanceUtil.getPreviousInstanceId(handle, instanceDto.getId());
+            boolean isFirstInstance = previousInstanceId == null;
+            boolean canDelete = ActivityInstanceUtil.computeCanDelete(
+                    activityDef.canDeleteInstances(),
+                    activityDef.getCanDeleteFirstInstance(),
+                    isFirstInstance);
+
             if (!activityDef.canDeleteInstances()) {
-                String msg = "Activity does not allow deleting instances";
-                LOG.warn(msg);
-                throw ResponseUtil.haltError(response, HttpStatus.SC_UNPROCESSABLE_ENTITY,
-                        new ApiError(ErrorCodes.OPERATION_NOT_ALLOWED, msg));
+                throwNotAllowed(response, "Activity does not allow deleting instances");
+            } else if (!canDelete) {
```

Comment:
`canDelete` means can delete an instance (even if it's not a first one). But error text says that first instance cannot be deleted..
What if `computeCanDelete` logic will be changed (for example, added a check of deleting a last instance): in that case the message could be incorrect.
May be to change it like:
```
boolean canDeleteFirstInstance = isFirstInstance + activityDef.getCanDeleteFirstInstance();
"Activity deletion not possible." + !canDeleteFirstInstance ?  " First instance cannot be deleted".
```
Or at least for more clear logic I'd suggest to set if-s like:
```
if (activityDef.canDeleteInstances()) {
   if (!canDelete) {
         throwNotAllowed(response, "Activity does not allow deleting the first instance");
   }
} else {
      throwNotAllowed(response, "Activity does not allow deleting instances");
}
```

Reply:
Good point! I'll change it.

Reply:
Should be better now!